### PR TITLE
Client Side investigation

### LINKED ISSUE #6353: Implement AMP for most read container
**Is your feature request related to a problem? Please describe.**
Blocked on: https://github.com/ampproject/amphtml/issues/27251

In order to render most read with fresh data, we have to use AMP HTML components, in the case of most read <amp-list> (https://amp.dev/documentation/components/amp-list/?format=websites) is sufficient.

We created a proof of concept for using amp-list to render most read on our AMP pages.
PR: https://github.com/bbc/simorgh/pull/5935 from this PR, we found a bug with amp-list which has been noted down on this issue: https://github.com/ampproject/amphtml/issues/27251
The issue is that amp-list does not responsively resize itself so this causes extra spacing to stay intact when a user changes their viewport from small to large. 

AMP components functionality are run on the client-side, this means that we are unable to write Javascript that allows us to manipulate the fetched data from amp-list. We are restricted to their mustache templating (basic for loops and if statements). The only piece of data we require to manipulate is the `rank` value from the endpoint: https://www.bbc.com/pidgin/mostread.json
The most read component in psammead would manipulate the `listIndex` value and return a native numeral. However, this is not a MUST requirement for AMP most read.

**Describe the solution you'd like**
- POC here: https://github.com/bbc/simorgh/pull/6440
- We would also have to update most read in the psammead component to be able to accept the `rank` value from the endpoint as a string.
- We should ensure most-read styling is still consistent with our AMP variation.
- When we are happy with the above, we will also have to create a new route for `bbc.com/$service/popular/read.amp`

**Describe alternatives you've considered**

**Testing notes**
[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [ ] This feature is expected to need manual testing.

**Additional context**
Note: most read is currently being used on Article, Front, popular ((/popular/read)), and STY page.

We conducted some initial investigations into different approaches for most read on AMP, which can be seen on this issue:https://github.com/bbc/simorgh/issues/4742

TL:DR:
- We tried using amp-script but this was not possible to use with ReactJS as the library was too large.
- We decided with the product team to cut out some requirements of most read for AMP. e.g to not render the timestamp (show timestamp of promo if it's greater than 60days) and the freshness requirement (Do not render most read when the last updated value is greater than 35min).

### PR DESCRIPTION
**Pull request by DarioR01**
Resolves #6353 

**Overall change:**
Continuing from the above investigation, we tried to implement most-read AMP in STY and Article for both WS and PS services.
Our solution is based on the AMP documentation for [using amp-script as a data source](https://amp.dev/documentation/components/amp-list/#using-amp-script-as-a-data-source)

### CODE REVIEW

**Comment by DarioR01** [comment]
### Introduction
We created a script that fetches data from the most read endpoint. The `<amp-list>` uses this script to fetch and display the most read data. As far as we know the data from the given endpoint is up to date, therefore within the 35 minutes requirement.

Limitation of this solution:
- We have not found a way to fix the height issue when switching between breakpoints described in the previous investigation.

- If the most read is in the second column and if the page is refreshed while the most read is in view, then it does not render properly. This issue does not seem to occur when placed at the bottom of the page. Therefore, not a limitation for mobile devices (refinement: maybe bad for tablet breakpoint?)

### Complications:
### `<amp-scrip>` : 
`<amp-script>` requires a meta tag in the header that holds the hash for the script. The hash is created by passing the innerHTML of the `<script>` tag to the hash function.
The hash needs to be updated if the script changes (make sure the format is exactly the same as the one rendered on the browser).

`<amp-script>` also requires a height and a width to work.

Finally, `<script>` needs the full path to the most-read endpoint for correctly fetching the data.
Example:
good: https://localhost:7080/mundo/mostread.json
bad: /mundo/mostread.json

### hash function:

```
const crypto = require('crypto');

export default function generateCSPHash(script) {
  const hash = crypto.createHash('sha384');
  const data = hash.update(script, 'utf8');
  return `sha384-${data
    .digest('base64')
    .replace(/=/g, '')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')}`;
}
```

 Because our fetch script dynamically changes depending on the service we render the most-read component, we had to calculate the value of the hash that we need in the meta tag.
From the <amp-script> [documentation](https://amp.dev/documentation/components/amp-script/#calculating-the-script-hash) we found a way of calculating the hash value we needed. 

When investigating this, we found complications with the `.digest('base64')` function, which does not run multiple times within the same hash object. Therefore we had to create a `crypto.createHash('sha384');` object everytime we run the function (aka: at every page load) by moving `const hash = crypto.createHash('sha384');` inside the function.
At the time of writing, we are unsure if this will have a negative impact on page performance.

Another complication we had found was that we needed to  add the following in package.json:

```
"browser": {
    "crypto": false
}
```
To fix a Webpack Breaking change:

Screenshot 2022-04-11 at 12 24 58

This will prevent Node.js from bundling the crypto dependency to serve on the browser.
We believe this is not the best way of dealing with this breaking change, and we should provide a polyfill for this library instead.

**Comment by DarioR01** [comment]
Above comment for our investigation outcome

File: src/app/containers/MostRead/Amp/index.jsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+const AmpMostRead = ({ endpoint, size, wrapper: Wrapper }) => {
+  const {
+    service,
+    script,
+    dir,
+    mostRead: { numberOfItems },
+  } = useContext(ServiceContext);
+
+  const onlyinnerscript = `const translations = { bengali: ['০', '১', '২', '৩', '৪', '৫', '৬', '৭', '৮', '৯', '১০'], burmese: ['၀', '၁', '၂', '၃', '၄', '၅', '၆', '၇', '၈', '၉', '၁၀'], nepali: ['०', '१', '२', '३', '४', '५', '६', '७', '८', '९', '१०'], EasternArabic: ['۰','۱','۲','۳','۴','۵','۶','۷','۸','۹','۱۰'], }; const WesternArabic = ['0','1','2','3','4','5','6','7','8','9','10']; function getRemoteData() { return fetch("${endpoint}") .then(resp => resp.json()) .then(resp => {resp.records.forEach((item, index) => item.rankTranslation = translations["${service}"]? translations["${service}"][[index+1]]: WesternArabic[index+1]); return resp;})} exportFunction('getRemoteData', getRemoteData);`;
```
Comment (andrewscfc) [inline src/app/containers/MostRead/Amp/index.jsx:22]: I think in the prod implementation we could clean this up a bit by building the script when the react component loads, this has a few advantages:
- If the translations change (likely due to a long standing bug/design change), they will be reflected in this string
- By only including the translations for the service we are on we decrease the page weight

Only a minor issue, mostly around readability in the future

File: src/app/containers/MostRead/Amp/index.jsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+          name="amp-script-src"
+          content={generateCSPHash(onlyinnerscript)}
+        />
+      </Helmet>
+
+      <amp-script
+        id="dataFunctions"
+        script="local-script"
+        width="1"
+        height="1"
```
Comment (andrewscfc) [inline src/app/containers/MostRead/Amp/index.jsx:56]: I can't really work out why the `amp-script` element needs a width and a height, the docs don't seem to shed much light 🤷 

It looks like the intent might be for the element to wrap the component that is being manipulated by the script, maybe is should surround `MostReadList` with a layout of `responsive`?
`<amp-script layout="responsive" id="dataFunctions" script="local-script"></amp-script>`

That could solve any potential a11y issues too

**Pull request closed without merging**